backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/ColumnarCachedBatchSerializer.scala

```diff
@@ -0,0 +1,303 @@
+      SparkSession.getActiveSession.orNull.sparkContext)
+    val numInputRows = metrics("numInputRows")
+    val numOutputBatches = metrics("numOutputBatches")
+    val convertTime = metrics("convertTime")
+    val numRows = conf.columnBatchSize
+    val rddColumnarBatch = input.mapPartitions {
+      it =>
+        RowToVeloxColumnarExec.toColumnarBatchIterator(
+          it,
+          localSchema,
+          numInputRows,
+          numOutputBatches,
+          convertTime,
+          numRows
+        )
+    }
+    convertColumnarBatchToCachedBatch(rddColumnarBatch, schema, storageLevel, conf)
+  }
+
+  override def convertCachedBatchToInternalRow(
+      input: RDD[CachedBatch],
+      cacheAttributes: Seq[Attribute],
+      selectedAttributes: Seq[Attribute],
+      conf: SQLConf): RDD[InternalRow] = {
+    if (!validateSchema(cacheAttributes)) {
+      // if we do not support this schema that means we are using row-based serializer,
+      // see `convertInternalRowToCachedBatch`, so fallback to vanilla Spark serializer
+      return rowBasedCachedBatchSerializer.convertCachedBatchToInternalRow(
+        input,
+        cacheAttributes,
+        selectedAttributes,
+        conf)
+    }
+
+    // note, these metrics are unused but just make `VeloxColumnarToRowExec` happy
+    val metrics = BackendsApiManager.getMetricsApiInstance.genColumnarToRowMetrics(
+      SparkSession.getActiveSession.orNull.sparkContext)
+    val numOutputRows = metrics("numOutputRows")
+    val numInputBatches = metrics("numInputBatches")
+    val convertTime = metrics("convertTime")
+    val rddColumnarBatch =
+      convertCachedBatchToColumnarBatch(input, cacheAttributes, selectedAttributes, conf)
+    rddColumnarBatch.mapPartitions {
+      it =>
+        VeloxColumnarToRowExec.toRowIterator(
+          it,
+          selectedAttributes,
+          numOutputRows,
+          numInputBatches,
+          convertTime
+        )
+    }
+  }
+
+  override def convertColumnarBatchToCachedBatch(
+      input: RDD[ColumnarBatch],
+      schema: Seq[Attribute],
+      storageLevel: StorageLevel,
+      conf: SQLConf): RDD[CachedBatch] = {
+    input.mapPartitions {
+      it =>
+        new Iterator[CachedBatch] {
+          override def hasNext: Boolean = it.hasNext
+
+          override def next(): CachedBatch = {
+            val batch = it.next()
+            val serializeResult = ColumnarBatchSerializerJniWrapper.INSTANCE.serialize(
+              ExecutionCtxs.contextInstance().getHandle,
+              Array(ColumnarBatches.getNativeHandle(batch)),
+              NativeMemoryManagers
+                .contextInstance("ColumnarCachedBatchSerializer")
+                .getNativeInstanceHandle
+            )
+            CachedColumnarBatch(
+              Math.toIntExact(serializeResult.getNumRows),
+              serializeResult.getSerialized.length,
+              serializeResult.getSerialized)
+          }
```

Comment:
```suggestion
CachedColumnarBatch(Math.toIntExact(serializeResult.getNumRows),
              serializeResult.getSerialized.length,
              Array(serialized col 0, serialized col 1, serialized col 2..))
```

Regarding serialized col x, it's serialized from a columnar batch that only includes a single column (with ordinal x) from the original columnar batch:
```
for (x <- (0 until batch.columns)) {
  batch for col x = ColumnarBatches.select(batch, Array(x)); // since only a "x" in input ordinal array, returned batch only contains one single column.
  serialized col x = ColumnarBatchSerializerJniWrapper.INSTANCE.serialize(ctx, batch for col x, mm);
}
```